File: python/src/semantic_retrieval/generator/retrieval_augmented_generation/vector_db_rag_completion_generator.py
from typing import Generic

from semantic_retrieval.common.types import P
from semantic_retrieval.data_store.vector_dbs.vector_db import (
    VectorDBBaseQuery,
    VectorDBQuery,
    VectorDBTextQuery,
)
from semantic_retrieval.generator.retrieval_augmented_generation.rag_completion_generator import (
    RAGCompletionGenerator,
    RAGCompletionGeneratorParams,
)
from semantic_retrieval.retrieval.vector_dbs.vector_db_document_retriever import (
    VectorDBDocumentRetriever,
)


class VectorDBRAGCompletionGeneratorParams(
    Generic[P], RAGCompletionGeneratorParams[P, VectorDBQuery]
):
    def __init__(
        self,
        retriever: VectorDBDocumentRetriever,
        retrieval_query: VectorDBBaseQuery,
    ):
        self.retriever = retriever
        self.retriever_query = retrieval_query


class VectorDBRAGCompletionGenerator(
    Generic[P], RAGCompletionGenerator[VectorDBDocumentRetriever, P]
):
    async def get_retrieval_query(  # type: ignore
        self, params: VectorDBRAGCompletionGeneratorParams[P]
    ) -> VectorDBTextQuery:
        raise NotImplementedError()

File: typescript/src/generator/retrieval-augmented-generation/vectorDBRAGCompletionGenerator.ts
import {
  VectorDBBaseQuery,
  VectorDBTextQuery,
} from "../../data-store/vector-DBs/vectorDB";
import { VectorDBDocumentRetriever } from "../../retrieval/vector-DBs/vectorDBDocumentRetriever";
import {
  CompletionModel,
  ModelResponse,
} from "../completion-models/completionModel";
import {
  RAGCompletionGenerator,
  RAGCompletionGeneratorParams,
} from "./ragCompletionGenerator";

export interface VectorDBRAGCompletionGeneratorParams
  extends RAGCompletionGeneratorParams<VectorDBDocumentRetriever> {
  retrievalQuery?: VectorDBBaseQuery;
}

export class VectorDBRAGCompletionGenerator<
  M extends CompletionModel<ModelResponse<M>>,
> extends RAGCompletionGenerator<M, VectorDBDocumentRetriever> {
  async getRetrievalQuery(
    params: VectorDBRAGCompletionGeneratorParams
  ): Promise<VectorDBTextQuery> {
    const { prompt, retrievalQuery } = params;
    const text = typeof prompt === "string" ? prompt : await prompt.toString();
    const topK = retrievalQuery?.topK ?? 3;
    const query = {
      ...params.retrievalQuery,
      topK,
      text,
    };
    await this.callbackManager?.runCallbacks({
      name: "onGetRAGCompletionRetrievalQuery",
      params,
      query,
    });
    return query;
  }
}
